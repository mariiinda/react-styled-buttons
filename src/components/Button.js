import React, { useEffect, useState } from "react";
import { css } from "@emotion/core";

import themes from "../theme/themes";

// CSS
const ButtonStyle = theme => css`
  position: relative;
  cursor: pointer;
  outline: none;
  margin: 0;
  padding: 20px;
  width: auto;
  text-align: center;
  display: inline-block;
  border: none;
  border-radius: 6px;
  background: ${theme.primaryColor};
  color: ${theme.backgroundColor};
  font-size: 16px;
  font-family: ${theme.primaryFont};
  line-height: inherit;
  font-weight: inherit;
  transform: translate3d(0, 0, 0);
  transition: 0.2s ease-in-out;

  &:hover {
    transform: translate3d(0, -3px, 0);
    box-shadow: 0 2px 5px rgba(0, 0, 0, 0.15);
  }

  &:active {
    background: ${theme.primaryShade5};
    transform: translate3d(0, 0, 0);
    box-shadow: 0 2px 5px rgba(0, 0, 0, 0.15);
  }

  &:focus {
    box-shadow: 0 0px 8px ${theme.primaryShade1},
      0 0px 8px ${theme.primaryShade1};
  }

  &:disabled {
    opacity: 0.6;
  }
`;

const SecondaryButtonStyle = theme => css`
  background: ${theme.secondaryColor};
  color: ${theme.backgroundColor};

  &:active {
    background: ${theme.secondaryShade5};
  }

  &:focus {
    box-shadow: 0 0px 8px ${theme.secondaryShade1},
      0 0px 8px ${theme.secondaryShade1};
  }
`;

const Accent1ButtonStyle = theme => css`
  background: ${theme.accentColor};
  color: ${theme.backgroundColor};

  &:active {
    background: ${theme.accentShade5};
  }

  &:focus {
    box-shadow: 0 0px 8px ${theme.accentShade1}, 0 0px 8px ${theme.accentShade1};
  }
`;

const Accent2ButtonStyle = theme => css`
  background: ${theme.accent2Color};
  color: ${theme.backgroundColor};

  &:active {
    background: ${theme.accent2Shade5};
  }

  &:focus {
    box-shadow: 0 0px 8px ${theme.accent2Shade1},
      0 0px 8px ${theme.accent2Shade1};
  }
`;

const Accent3ButtonStyle = theme => css`
  background: ${theme.accent3Color};
  color: ${theme.backgroundColor};

  &:active {
    background: ${theme.accent3Shade5};
  }

  &:focus {
    box-shadow: 0 0px 8px ${theme.accent3Shade1},
      0 0px 8px ${theme.accent3Shade1};
  }
`;

const NeutralButtonStyle = theme => css`
  background: ${theme.neutralColor};
  color: ${theme.backgroundColor};

  &:active {
    background: ${theme.neutralColor5};
  }

  &:focus {
    box-shadow: 0 0px 8px ${theme.neutralColor4},
      0 0px 8px ${theme.neutralColor4};
  }
`;

const SmallButtonStyle = css`
  padding: 4px 12px;
  font-size: 14px;
`;

const MediumButtonStyle = css`
  padding: 10px;
`;

const composeStyles = ({ themeObject, variant, size }) => {
  const cssStyles = [ButtonStyle(themeObject)];
  const variantCases = {
    secondary: () => cssStyles.push(SecondaryButtonStyle(themeObject)),
    accent1: () => cssStyles.push(Accent1ButtonStyle(themeObject)),
    accent2: () => cssStyles.push(Accent2ButtonStyle(themeObject)),
    accent3: () => cssStyles.push(Accent3ButtonStyle(themeObject)),
    neutral: () => cssStyles.push(NeutralButtonStyle(themeObject))
  };
  if (variantCases[variant]) {
    variantCases[variant]();
  }
  const sizeCases = {
    small: () => cssStyles.push(SmallButtonStyle),
    medium: () => cssStyles.push(MediumButtonStyle)
  };
  if (sizeCases[size]) {
    sizeCases[size]();
  }
  return cssStyles;
};

function Button({ as: Element, id, onClick, variant, size, theme, ...props }) {
  const [themeObject, setThemeObject] = useState(themes[theme]);
  useEffect(() => {
    if (theme === "light" || theme === "dark") {
      console.log({ t: themes[theme], theme });
      setThemeObject(themes[theme]);
    }
  }, [theme]);
  const composedStyles = composeStyles({ themeObject, variant, size });
  const type = Element === "button" && !props.type ? "button" : null;

  console.log({ themeObject });

  return (
    <Element
      //css={composedStyles}
      id={id}
      data-testid={id}
      onClick={onClick}
      type={type}
      {...props}
    >
      <span>{props.children}</span>
    </Element>
  );
}

Button.defaultProps = {
  as: "button",
  id: "",
  children: null,
  disabled: false,
  onClick: () => {},
  theme: "light",
  variant: "primary",
  size: "large"
};

// enumarables: todo add to proptypes
// variant: ["primary", "secondary", "accent1", "accent2", "accent3", "neutral"]
// size: ["large", "medium", "small"]
// theme ["light","dark"]

export default Button;
