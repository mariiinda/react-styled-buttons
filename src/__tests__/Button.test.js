import React from "react";
import "@testing-library/react/cleanup-after-each";
import "@testing-library/jest-dom/extend-expect";
import { render, cleanup } from "@testing-library/react";

import Button from "../components/Button";

afterEach(cleanup);

test("visible focus state", () => {
  const { getByTestId } = render(<Button id="my-button">Click me</Button>);
  getByTestId("my-button").focus();
  expect(getByTestId("my-button")).toHaveFocus();
});

test("renders as disbled", () => {
  const { getByTestId } = render(
    <Button id="my-button" disabled>
      Click me
    </Button>
  );
  expect(getByTestId("my-button")).toBeDisabled();
});

test("render as anchor element", () => {
  const element = "a";
  const { container } = render(
    <Button id="my-button" as="a">
      Click me
    </Button>
  );
  const anchorEl = container.querySelector(element);
  expect(anchorEl).toBeInTheDocument();
});

test("renders dark theme", () => {
  const theme = "dark";
  const { getByTestId } = render(
    <Button id="my-button" theme={theme}>
      Click me
    </Button>
  );
  expect(getByTestId("my-button")).toMatchSnapshot();
});

test("renders secondary variant", () => {
  const variant = "secondary";
  const { getByTestId } = render(
    <Button id="my-button" variant={variant}>
      Click me
    </Button>
  );
  expect(getByTestId("my-button")).toMatchSnapshot();
});

test("renders accent1 variant", () => {
  const variant = "accent1";
  const { getByTestId } = render(
    <Button id="my-button" variant={variant}>
      Click me
    </Button>
  );
  expect(getByTestId("my-button")).toMatchSnapshot();
});
test("renders accent2 variant", () => {
  const variant = "accent2";
  const { getByTestId } = render(
    <Button id="my-button" variant={variant}>
      Click me
    </Button>
  );
  expect(getByTestId("my-button")).toMatchSnapshot();
});

test("renders neutral variant", () => {
  const variant = "neutral";
  const { getByTestId } = render(
    <Button id="my-button" variant={variant}>
      Click me
    </Button>
  );
  expect(getByTestId("my-button")).toMatchSnapshot();
});

test("renders medium size", () => {
  const size = "medium";
  const { getByTestId } = render(
    <Button id="my-button" size={size}>
      Click me
    </Button>
  );
  expect(getByTestId("my-button")).toMatchSnapshot();
});

test("renders small size", () => {
  const size = "small";
  const { getByTestId } = render(
    <Button id="my-button" size={size}>
      Click me
    </Button>
  );
  expect(getByTestId("my-button")).toMatchSnapshot();
});

test("renders correctly", () => {
  const testChildren = "Click me";
  const { container, getByText } = render(
    <Button id="my-button">{testChildren}</Button>
  );
  expect(getByText(testChildren)).toBeInTheDocument();

  expect(container.firstChild).toMatchSnapshot();
});
