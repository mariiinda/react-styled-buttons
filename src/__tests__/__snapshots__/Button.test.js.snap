// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders accent1 variant 1`] = `
.emotion-0 {
  position: relative;
  cursor: pointer;
  outline: none;
  margin: 0;
  padding: 20px;
  width: auto;
  text-align: center;
  display: inline-block;
  border: none;
  border-radius: 6px;
  background: #1971c2;
  color: #ffffff;
  font-size: 16px;
  font-family: 'Fira Sans',Arial,Helvetica,sans-serif;
  line-height: inherit;
  font-weight: inherit;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  background: #0b7285;
  color: #ffffff;
}

.emotion-0:hover {
  -webkit-transform: translate3d(0,-3px,0);
  -ms-transform: translate3d(0,-3px,0);
  transform: translate3d(0,-3px,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:active {
  background: #1864ab;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:focus {
  box-shadow: 0 0px 8px #74c0fc,0 0px 8px #74c0fc;
}

.emotion-0:disabled {
  opacity: 0.6;
}

.emotion-0:active {
  background: #0c8599;
}

.emotion-0:focus {
  box-shadow: 0 0px 8px #66d9e8,0 0px 8px #66d9e8;
}

<button
  class="emotion-0"
  data-testid="my-button"
  id="my-button"
  type="button"
>
  <span>
    Click me
  </span>
</button>
`;

exports[`renders accent2 variant 1`] = `
.emotion-0 {
  position: relative;
  cursor: pointer;
  outline: none;
  margin: 0;
  padding: 20px;
  width: auto;
  text-align: center;
  display: inline-block;
  border: none;
  border-radius: 6px;
  background: #1971c2;
  color: #ffffff;
  font-size: 16px;
  font-family: 'Fira Sans',Arial,Helvetica,sans-serif;
  line-height: inherit;
  font-weight: inherit;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  background: #087f5b;
  color: #ffffff;
}

.emotion-0:hover {
  -webkit-transform: translate3d(0,-3px,0);
  -ms-transform: translate3d(0,-3px,0);
  transform: translate3d(0,-3px,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:active {
  background: #1864ab;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:focus {
  box-shadow: 0 0px 8px #74c0fc,0 0px 8px #74c0fc;
}

.emotion-0:disabled {
  opacity: 0.6;
}

.emotion-0:active {
  background: #099268;
}

.emotion-0:focus {
  box-shadow: 0 0px 8px #63e6be,0 0px 8px #63e6be;
}

<button
  class="emotion-0"
  data-testid="my-button"
  id="my-button"
  type="button"
>
  <span>
    Click me
  </span>
</button>
`;

exports[`renders correctly 1`] = `
.emotion-0 {
  position: relative;
  cursor: pointer;
  outline: none;
  margin: 0;
  padding: 20px;
  width: auto;
  text-align: center;
  display: inline-block;
  border: none;
  border-radius: 6px;
  background: #1971c2;
  color: #ffffff;
  font-size: 16px;
  font-family: 'Fira Sans',Arial,Helvetica,sans-serif;
  line-height: inherit;
  font-weight: inherit;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
}

.emotion-0:hover {
  -webkit-transform: translate3d(0,-3px,0);
  -ms-transform: translate3d(0,-3px,0);
  transform: translate3d(0,-3px,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:active {
  background: #1864ab;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:focus {
  box-shadow: 0 0px 8px #74c0fc,0 0px 8px #74c0fc;
}

.emotion-0:disabled {
  opacity: 0.6;
}

<button
  class="emotion-0"
  data-testid="my-button"
  id="my-button"
  type="button"
>
  <span>
    Click me
  </span>
</button>
`;

exports[`renders dark theme 1`] = `
.emotion-0 {
  position: relative;
  cursor: pointer;
  outline: none;
  margin: 0;
  padding: 20px;
  width: auto;
  text-align: center;
  display: inline-block;
  border: none;
  border-radius: 6px;
  background: #4dabf7;
  color: #000000;
  font-size: 16px;
  font-family: 'Fira Sans',Arial,Helvetica,sans-serif;
  line-height: inherit;
  font-weight: inherit;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
}

.emotion-0:hover {
  -webkit-transform: translate3d(0,-3px,0);
  -ms-transform: translate3d(0,-3px,0);
  transform: translate3d(0,-3px,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:active {
  background: #339af0;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:focus {
  box-shadow: 0 0px 8px #228be6,0 0px 8px #228be6;
}

.emotion-0:disabled {
  opacity: 0.6;
}

<button
  class="emotion-0"
  data-testid="my-button"
  id="my-button"
  type="button"
>
  <span>
    Click me
  </span>
</button>
`;

exports[`renders medium size 1`] = `
.emotion-0 {
  position: relative;
  cursor: pointer;
  outline: none;
  margin: 0;
  padding: 20px;
  width: auto;
  text-align: center;
  display: inline-block;
  border: none;
  border-radius: 6px;
  background: #1971c2;
  color: #ffffff;
  font-size: 16px;
  font-family: 'Fira Sans',Arial,Helvetica,sans-serif;
  line-height: inherit;
  font-weight: inherit;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  padding: 10px;
}

.emotion-0:hover {
  -webkit-transform: translate3d(0,-3px,0);
  -ms-transform: translate3d(0,-3px,0);
  transform: translate3d(0,-3px,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:active {
  background: #1864ab;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:focus {
  box-shadow: 0 0px 8px #74c0fc,0 0px 8px #74c0fc;
}

.emotion-0:disabled {
  opacity: 0.6;
}

<button
  class="emotion-0"
  data-testid="my-button"
  id="my-button"
  type="button"
>
  <span>
    Click me
  </span>
</button>
`;

exports[`renders neutral variant 1`] = `
.emotion-0 {
  position: relative;
  cursor: pointer;
  outline: none;
  margin: 0;
  padding: 20px;
  width: auto;
  text-align: center;
  display: inline-block;
  border: none;
  border-radius: 6px;
  background: #1971c2;
  color: #ffffff;
  font-size: 16px;
  font-family: 'Fira Sans',Arial,Helvetica,sans-serif;
  line-height: inherit;
  font-weight: inherit;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  background: #495057;
  color: #ffffff;
}

.emotion-0:hover {
  -webkit-transform: translate3d(0,-3px,0);
  -ms-transform: translate3d(0,-3px,0);
  transform: translate3d(0,-3px,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:active {
  background: #1864ab;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:focus {
  box-shadow: 0 0px 8px #74c0fc,0 0px 8px #74c0fc;
}

.emotion-0:disabled {
  opacity: 0.6;
}

.emotion-0:active {
  background: #343a40;
}

.emotion-0:focus {
  box-shadow: 0 0px 8px #868e96,0 0px 8px #868e96;
}

<button
  class="emotion-0"
  data-testid="my-button"
  id="my-button"
  type="button"
>
  <span>
    Click me
  </span>
</button>
`;

exports[`renders secondary variant 1`] = `
.emotion-0 {
  position: relative;
  cursor: pointer;
  outline: none;
  margin: 0;
  padding: 20px;
  width: auto;
  text-align: center;
  display: inline-block;
  border: none;
  border-radius: 6px;
  background: #1971c2;
  color: #ffffff;
  font-size: 16px;
  font-family: 'Fira Sans',Arial,Helvetica,sans-serif;
  line-height: inherit;
  font-weight: inherit;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  background: #6741d9;
  color: #ffffff;
}

.emotion-0:hover {
  -webkit-transform: translate3d(0,-3px,0);
  -ms-transform: translate3d(0,-3px,0);
  transform: translate3d(0,-3px,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:active {
  background: #1864ab;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:focus {
  box-shadow: 0 0px 8px #74c0fc,0 0px 8px #74c0fc;
}

.emotion-0:disabled {
  opacity: 0.6;
}

.emotion-0:active {
  background: #5f3dc4;
}

.emotion-0:focus {
  box-shadow: 0 0px 8px #b197fc,0 0px 8px #b197fc;
}

<button
  class="emotion-0"
  data-testid="my-button"
  id="my-button"
  type="button"
>
  <span>
    Click me
  </span>
</button>
`;

exports[`renders small size 1`] = `
.emotion-0 {
  position: relative;
  cursor: pointer;
  outline: none;
  margin: 0;
  padding: 20px;
  width: auto;
  text-align: center;
  display: inline-block;
  border: none;
  border-radius: 6px;
  background: #1971c2;
  color: #ffffff;
  font-size: 16px;
  font-family: 'Fira Sans',Arial,Helvetica,sans-serif;
  line-height: inherit;
  font-weight: inherit;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  padding: 4px 12px;
  font-size: 14px;
}

.emotion-0:hover {
  -webkit-transform: translate3d(0,-3px,0);
  -ms-transform: translate3d(0,-3px,0);
  transform: translate3d(0,-3px,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:active {
  background: #1864ab;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
  box-shadow: 0 2px 5px rgba(0,0,0,0.15);
}

.emotion-0:focus {
  box-shadow: 0 0px 8px #74c0fc,0 0px 8px #74c0fc;
}

.emotion-0:disabled {
  opacity: 0.6;
}

<button
  class="emotion-0"
  data-testid="my-button"
  id="my-button"
  type="button"
>
  <span>
    Click me
  </span>
</button>
`;
